url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwin/geneticOperation/ShrinkMutation.java

```diff
@@ -0,0 +1,59 @@
+package darwin.geneticOperation;
+
+import java.util.ArrayList;
+import java.util.List;
+import darwin.initialization.NoAvailableNodeException;
+import darwin.nodeFilter.All;
+import darwin.nodeFilter.Functions;
+import darwin.nodeFilter.Terminals;
+import darwin.population.Individual;
+import darwin.population.Node;
+import darwin.population.TypeMismatchException;
+import darwin.problemObject.NodeFactory;
+
+public class ShrinkMutation implements GeneticOperation {
+
+	private NodeFactory nodeFactory;
+	
+	public ShrinkMutation(NodeFactory nodeFactory) {
+		this.nodeFactory = nodeFactory;
+	}
+	
+	@Override
+	public int getNumberInputs() {
+		return 1;
+	}
+
+	@Override
+	public Individual[] performOperation(Individual[] inputs, int maxSize, boolean function) throws UnableToPerformOperationException {
+		Individual father = inputs[0];
+		
+		List<Node> yList = new ArrayList<Node>();
+		if(function)
+			yList = father.filterTrees(new Functions(Object.class));
+		if(yList.size() == 0)
+			yList = father.filterTrees(new All());
+		
+		int random = (int) (Math.random() * yList.size());
+		Node yChromosome = yList.get(random);
+		
+		Class<?> expectedType = yChromosome.getExpectedType();
+		
+		Node replacement;
+		try {
+			replacement = nodeFactory.getNode(new Terminals(expectedType));
+		} catch (NoAvailableNodeException e1) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)